proof/refine/InvariantUpdates_H.thy

```diff
@@ -0,0 +1,457 @@
+
+(* arch-specific typ_at_lifts in Arch *)
+lemmas gen_typ_at_lifts =
+         typ_at_lift_tcb' typ_at_lift_ep' typ_at_lift_ntfn' typ_at_lift_cte' typ_at_lift_cte_at'
+         typ_at_lift_valid_untyped' typ_at_lift_valid_cap' valid_bound_tcb_lift
+         valid_arch_tcb_lift'
+
+(* these depend on interpretations in ArchInvLemmas_H *)
+context pspace_update_eq'
+begin
+
+lemma valid_space_update [iff]:
+  "valid_pspace' (f s) = valid_pspace' s"
+  by (fastforce simp: valid_pspace' pspace)
+
+lemma cte_wp_at_update [iff]:
+  "cte_wp_at' P p (f s) = cte_wp_at' P p s"
+  by (fastforce intro: cte_wp_at'_pspaceI simp: pspace)
+
+lemma ex_nonz_cap_to_eq'[iff]:
+  "ex_nonz_cap_to' p (f s) = ex_nonz_cap_to' p s"
+  by (simp add: ex_nonz_cap_to'_def)
+
+lemma iflive_update [iff]:
+  "if_live_then_nonz_cap' (f s) = if_live_then_nonz_cap' s"
+  by (simp add: if_live_then_nonz_cap'_def ex_nonz_cap_to'_def)
+
+lemma valid_objs_update [iff]:
+  "valid_objs' (f s) = valid_objs' s"
+  apply (simp add: valid_objs'_def pspace)
+  apply (fastforce intro: valid_obj'_pspaceI simp: pspace)
+  done
+
+lemma valid_cap_update [iff]:
+  "(f s) \<turnstile>' c = s \<turnstile>' c"
+  by (auto intro: valid_cap'_pspaceI simp: pspace)
+
+(* exports from Arch locale version which are safe for generic use *)
+interpretation Arch_pspace_update_eq' ..
+
+lemmas pspace_in_kernel_mappings_update'[iff] = pspace_in_kernel_mappings_update'
+
+end
+
+context p_arch_idle_update_eq'
+begin
+
+lemma ifunsafe_update [iff]:
+  "if_unsafe_then_cap' (f s) = if_unsafe_then_cap' s"
+  by (simp add: if_unsafe_then_cap'_def ex_cte_cap_to'_def int_nd)
+
+(* exports from Arch locale version (safe for generic use) *)
+interpretation Arch_p_arch_idle_update_eq' ..
+
+lemmas valid_arch_state_update'[iff] = valid_arch_state_update'
+
+end
+
+locale InvariantUpdates_H =
+  assumes valid_arch_state'_interrupt[simp]:
+    "\<And>f s. valid_arch_state' (ksInterruptState_update f s) = valid_arch_state' s"
+
+(* FIXME: use locales to shorten this work *)
+
+lemma ps_clear_domE[elim?]:
+  "\<lbrakk> ps_clear x n s; dom (ksPSpace s) = dom (ksPSpace s') \<rbrakk> \<Longrightarrow> ps_clear x n s'"
+  by (simp add: ps_clear_def)
+
+lemma ps_clear_upd:
+  "ksPSpace s y = Some v \<Longrightarrow>
+    ps_clear x n (ksPSpace_update (\<lambda>a. (ksPSpace s)(y \<mapsto> v')) s') = ps_clear x n s"
+  by (rule iffI | clarsimp elim!: ps_clear_domE | fastforce)+
+
+lemmas ps_clear_updE[elim] = iffD2[OF ps_clear_upd, rotated]
+
+lemma ct_not_inQ_ksMachineState_update[simp]:
+  "ct_not_inQ (ksMachineState_update f s) = ct_not_inQ s"
+  by (simp add: ct_not_inQ_def)
+
+lemma ct_in_current_domain_ksMachineState[simp]:
+  "ct_idle_or_in_cur_domain' (ksMachineState_update p s) = ct_idle_or_in_cur_domain' s"
+  by (simp add: ct_idle_or_in_cur_domain'_def tcb_in_cur_domain'_def)
+
+(* FIXME arch-split: export? *)
+lemma invs'_machine:
```

Comment:
I don't think I understand the comment. We're in a generic file at the top level, so where would we export to?

Reply:
I think it's a leftover from when it was in the Arch locale, forgot to remove comment when I moved it.

- [ ] remove `FIXME arch-split: export?` comments here and for `invs_no_cicd'_machine`